### DISCUSSION TITLE
Confusing documentation on delete cascades and many-to-many relationships.

### DISCUSSION THREAD

**Discussion by mjpieters** [sqlalchemy/sqlalchemy#10614] [Ideas]
### Describe the bug

In the documentation on cascades, the [*Using delete cascade with many-to-many relationships* section](https://docs.sqlalchemy.org/en/20/orm/cascades.html#using-delete-cascade-with-many-to-many-relationships) contains this text:

> The `cascade="all, delete"` option works equally well with a many-to-many relationship, one that uses [relationship.secondary](https://docs.sqlalchemy.org/en/20/orm/relationship_api.html#sqlalchemy.orm.relationship.params.secondary) to indicate an association table.

The whole section then repeatedly references `cascade="all, delete"`. Given that `all` already includes `delete`,  I found this text confusing. Why not simply talk about using `delete` in  `cascade`, mentioning that `all` includes `delete`?

Next, I think it needs to be made explicit on what to use when you *don't* want objects on the opposite side of the many-to-many relationship to be deleted. This would be the case for *most* many-to-many releationships, given that the opposite object can be part of multiple such relationships. The use of `cascade="all, delete"` in this section might give the impression that using `cascade="all"` without the `delete` would be sufficient to prevent this. I'd prefer it if the documentation were to make this more explicit  (e.g. use, at most, `cascade="save-update, merge, refresh-expire, expunge"`).

I've included a test script that verifies that `cascade="save-update, merge, refresh-expire, expunge"` is sufficient to clear association table rows (but the default cascade would suffice).

### Optional link from https://docs.sqlalchemy.org which documents the behavior that is expected

https://docs.sqlalchemy.org/en/20/orm/cascades.html#using-delete-cascade-with-many-to-many-relationships

### SQLAlchemy Version in Use

2.0.23

### To Reproduce

```python
from __future__ import annotations

import sqlalchemy as sa
from sqlalchemy.orm import (
    DeclarativeBase,
    Mapped,
    mapped_column,
    relationship,
    sessionmaker,
)

engine = sa.create_engine("sqlite://", echo=True)

class Base(DeclarativeBase):
    pass

# Many-to-Many relationship between Foo and Bar objects

foo_bar = sa.Table(
    "foo_bar",
    Base.metadata,
    sa.Column("foo_id", sa.ForeignKey("foo.id"), primary_key=True),
    sa.Column("bar_id", sa.ForeignKey("bar.id"), primary_key=True),
)

class Foo(Base):
    __tablename__ = "foo"

    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str]

    bars: Mapped[list[Bar]] = relationship(
        secondary=foo_bar, cascade="save-update, merge, refresh-expire, expunge"
    )

# Many-to-Many relationship between FooAll and Bar objects

foo_all_bar = sa.Table(
    "foo_all_bar",
    Base.metadata,
    sa.Column("foo_id", sa.ForeignKey("foo_all.id"), primary_key=True),
    sa.Column("bar_id", sa.ForeignKey("bar.id"), primary_key=True),
)


# Same as Foo except this uses the "all" cascade option, adding "delete" to the options.
class FooAll(Base):
    __tablename__ = "foo_all"

    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str]

    bars: Mapped[list[Bar]] = relationship(secondary=foo_all_bar, cascade="all")

class Bar(Base):
    __tablename__ = "bar"

    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str]


Base.metadata.create_all(engine)
Session = sessionmaker(bind=engine)
session = Session()

foo1 = Foo(name="Foo 1")
foo2 = Foo(name="Foo 2")
foo_all1 = FooAll(name="Foo All 1")
foo_all2 = FooAll(name="Foo All 2")
bar1 = Bar(name="Bar 1")
bar2 = Bar(name="Bar 2")

foo1.bars = [bar1, bar2]
foo2.bars = [bar2, bar1]
foo_all1.bars = [bar1, bar2]
foo_all2.bars = [bar2, bar1]

session.add_all([foo1, foo2, foo_all1, foo_all2])
session.commit()

# There are 4 associations between Foo and Bar rows now
assert session.scalar(sa.select(sa.func.count(None)).select_from(foo_bar)) == 4

# Verify that deleting foo does *not* mark bar1 ar bar2 for deletion
session.delete(foo1)
assert foo1 in session.deleted
assert bar1 not in session.deleted
assert bar2 not in session.deleted
session.commit()

# 2 association table rows are now gone
assert session.scalar(sa.select(sa.func.count(None)).select_from(foo_bar)) == 2

# Deleting FooAll objects, on the other hand, causes Bar objects to be deleted too
# because the `delete` cascade is in effect. It doesn't matter that foo2 and foo_all2
# also relate to these objects.
session.delete(foo_all1)
assert foo_all1 in session.deleted
assert bar1 in session.deleted
assert bar2 in session.deleted
```

**Comment by zzzeek** [comment]
im on the road right now but the short answer is these cascades and conventions were all copied from hibernate a very long time ago.     if we want to revisit how they are described it might better to replace the whole thing with a more modern datastructure.

**Comment by zzzeek** [comment]
this is all discussion stuff until actionable items are taken so moving this to "ideas"

**Comment by mjpieters** [comment]
Sure, I'm fine with that. I noticed that `cascade="all, delete"` is used in more places, so this needs a broader audit either way.

**Comment by zzzeek** [comment]
> 
> The whole section then repeatedly references `cascade="all, delete"`. Given that `all` already includes `delete`, I found this text confusing. Why not simply talk about using `delete` in `cascade`, mentioning that `all` includes `delete`?

I've never been happy with this either.  however, as it is now, if you do this:

    cascade='delete'

then you lose all the other cascades, and if you do this:

    cascade='all'

the word "delete" is not even there, and this seems misleading also.  since it *doesnt* include `delete-orphan`.   It's very unusual someone would want "delete" and not "delete-orphan".   A superior approach here would likely use some term like "is_dependent" to indicate "delete, delete-orphan" and them some lesser used option like "is_dependent,ignore_orphans" if someone really wanted orphans to exist but still cascade deletes.



> 
> Next, I think it needs to be made explicit on what to use when you _don't_ want objects on the opposite side of the many-to-many relationship to be deleted. This would be the case for _most_ many-to-many releationships, given that the opposite object can be part of multiple such relationships. The use of `cascade="all, delete"` in this section might give the impression that using `cascade="all"` without the `delete` would be sufficient to prevent this. I'd prefer it if the documentation were to make this more explicit (e.g. use, at most, `cascade="save-update, merge, refresh-expire, expunge"`).

I think the doc section is pretty clear that if you *dont* do anything with cascade, the association rows are deleted:

> When a parent object is deleted, and therefore de-associated with its related objects, the unit of work process will normally delete rows from the association table, but leave the related objects intact. When combined with cascade="all, delete", additional DELETE statements will take place for the child rows themselves.

As far as telling people, "in order to allow association rows only to be deleted, set cascade="save-update, merge, refresh-expire, expunge", I am not in favor of that.  Listing out all the specific cascades is unnecessary and I dont want people ever having to list them out like this, that's why there's a default here and also if they want to add "delete" and/or "delete-orphan" there's an "all" option so that you dont lose that list.

The reality is that nobody ever changes any of these other cascade settings.  so the current norm is, "omit cascade entirely if you dont want a deletion of the endpoint object (secondary table or not), use "all, delete" for the full suite of deletion mechanics".   That's the coarse-grained level at which these options are currently used in practice.     so you will not see people saying `cascade="save-update, merge, refresh-expire, expunge"` and I would not want to suggest people should be doing that, that would be misleading and error prone and also make it harder for us to put a better API here.

Better API would include these features:

a. it would use pep-484 typing features so that the available options can be seen in IDEs
b. it would make clear which cascades are defaults and are always on, and would never require writing out each one
c. it would focus mainly on the real world selection people are making which is if object B is dependent on being attached to an object A, that is "delete, delete-orphan" behavior, with an option to turn off the "delete-orphan" part
d. it should likely unify with the [passive_deletes](https://docs.sqlalchemy.org/en/20/orm/relationship_api.html#sqlalchemy.orm.relationship.params.passive_deletes) set of options which is also pretty bad right now as it's a three-state not-really-boolean as well as the [passive_updates](https://docs.sqlalchemy.org/en/20/orm/relationship_api.html#sqlalchemy.orm.relationship.params.passive_updates) parameter.

constructs we can use here to allow for clearer use with typing support include things like [typed dicts](https://peps.python.org/pep-0589/) , named tuples, and enums.

**Reply by CaselIT** [reply]
> constructs we can use here to allow for clearer use with typing support include things like [typed dicts](https://peps.python.org/pep-0589/) , named tuples, and enums.

As an initial opinion I think the best option here is a Flag enum, that would look like this
```py
import enum
class CascadeOptions(enum.Flag):
    SAVE_UPDATE = enum.auto()
    MERGE = enum.auto()
    REFRESH_EXPIRE=enum.auto()
    EXPUNGE=enum.auto()

    DELETE = enum.auto()
    DELETE_ORPHAN = enum.auto()
    
    ALL = SAVE_UPDATE | MERGE | REFRESH_EXPIRE | EXPUNGE | DELETE  # ALL could be renamed to DEFAULT
```

Regarding also unifying these with the passive deletes, it may be possible, but we would have to look how to make the api convenient, since with flag enum you can only add options, not remove them (AFAIK).
So the default for passie-delete could be a "passive_delete_default" that gets superseded if another option is specified or nothing

**Reply by zzzeek** [reply]
> but we would have to look how to make the api convenient, since with flag enum you can only add options,

so maybe flag enum is not appropriate here?   how about some combination of flags / typed dict?

**Reply by zzzeek** [reply]
it would be good if people could have a simple API that is, "what do i do when the parent is deleted?  do i set the column to NULL?  do I delete the row? or do i do nothing?"  that's why passive_deletes seems like it should integrate with "delete cascade"

**Reply by zzzeek** [reply]
also, I dont think people should have to add imports in order to set "delete" cascade. that's why I liked TypedDict here.   making people import an Enum seems too heavy handed.

**Comment by PaleNeutron** [comment]
Sorry for re-opening the old discussion. But I'm not sure about which is the best config for user who want to cascade delete only rows in associated table not the related object.

In this example: https://docs.sqlalchemy.org/en/20/orm/cascades.html#using-delete-cascade-with-many-to-many-relationships, when a Parent object is marked for deletion, keep all children and just delete rows in `association_table`.

I think we need another example in the document for this condition. Maybe a `teacher` and `student` m2m relationship? Delete any one should not affect others.

**Reply by zzzeek** [reply]
if you empty a many-to-many collection , during flush the association rows are deleted and the related object is not deleted.  that is, you would not use delete-cascade here.

**Reply by PaleNeutron** [reply]
@zzzeek  If I delete a teacher without empty his's student. Will association rows automatically deleted?

**Reply by zzzeek** [reply]
always, yes

**Comment by panda-byte** [comment]
I also found this confusing, and I think the documentation would really benefit from an explanation of of why `all, delete` is the idiomatic usage and if (or why) it is recommended for users to do the same. It could be just one sentence. Something like:

"Although `all` already includes `delete`, the examples use `all, delete` explicitly — both for clarity and to emphasize the contrast with `all, delete-orphan`."

If this thing has tripped up two people already, I think many more readers of the docs were probably confused by it. Personally, I wasn't even sure that `all` did really include `delete`, because there were around 10 mentions of `all, delete` but only one mention of `delete` being included in `all`, so I ended up looking into the source code to make sure.

**Reply by zzzeek** [reply]
we can accept PRs in this area
